### PR TITLE
NATS span exporter

### PR DESCRIPTION
**Pull request by darwin67**
## Description

Add a NATS span exporter that can be used to publish otel spans to NATS directly.

- Add protobuf schema for the message schema
- Move span.go into `run` pkg to resolve circular dependency (also makes more sense)

## Type of change (choose one)
- [ ] Chore (refactors, upgrades, etc.)
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Security fix (non-breaking change that fixes a potential vulnerability)
- [ ] Docs
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Checklist
- [ ] I've linked any associated issues to this PR.
- [ ] I've tested my own changes.

*[Check our Pull Request Guidelines](https://github.com/inngest/inngest/blob/main/docs/PULL_REQUEST_GUIDELINES.md)*

### CODE REVIEW

File: pkg/pubsub/broker/nats.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func NewNATSConnector(ctx context.Context, opts NatsConnOpt) (*NatsConnector, error) {
+	host, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving hostname: %w", err)
+	}
+	name := fmt.Sprintf("%s-%s", opts.Name, host)
+
+	urls := "nats://127.0.0.1:4222"
```
Comment (djfarrelly) [inline pkg/pubsub/broker/nats.go:58]: suggestion: could be good as a const

File: pkg/pubsub/broker/nats.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	if cons, ok := c.consumers[key]; ok {
+		return cons, nil
+	}
+
+	return nil, fmt.Errorf("no consumer named '%s' initialized for stream: '%s'", consumer, stream)
+}
+
+// Shutdown drains the connection to flush data in buffer before fully closing the
+// connection
+func (c *NatsConnector) Shutdown(ctx context.Context) error {
```
Comment (djfarrelly) [inline pkg/pubsub/broker/nats.go:145]: **question**: Do we have to do anything to clean up the consumers? Like `DeleteConsumer`?

File: proto/run/v2/run.proto [hunk 3]
```diff
@@ -49,6 +53,9 @@ enum SpanStepOp {
   INVOKE = 1;
   SLEEP = 2;
   WAIT_FOR_EVENT = 3;
+  STEP = 4;
```
Comment (djfarrelly) [inline proto/run/v2/run.proto:56]: **question**: What is the difference between `STEP` and `RUN`?

File: cmd/commands/dev.go [hunk 4]
```diff
@@ -88,10 +89,14 @@ func doDev(cmd *cobra.Command, args []string) {
...
-		ServiceName:   "devserver",
-		Type:          itrace.TracerTypeOTLPHTTP,
+		ServiceName: "devserver",
+		Type:        itrace.TracerTypeOTLPHTTP,
+		// Type:          itrace.TracerTypeNATS,
 		TraceEndpoint: "localhost:8288",
 		TraceURLPath:  "/dev/traces",
+		NATS: &exporters.NatsExporterOpts{
+			Subjects: []string{"inngest.run.spans"},
+		},
```
Comment (djfarrelly) [inline cmd/commands/dev.go:99]: **question**: when this is added, will it swap out to an in-memory pubsub via the gocloud package when running the dev server? Maybe I'm just not sure how this was working before with the previous tracer anyway.

**Review by djfarrelly — APPROVED** [review]
Code looks good, left some non-blocking questions

File: pkg/pubsub/broker/nats.go (hunk 2 above)
Comment (darwin67) [reply pkg/pubsub/broker/nats.go:145]: No, it shouldn't be clean up.
A consumer is a nats resource, created on the nats server side. we merely authenticate as a consumer to pull messages to work on.

File: proto/run/v2/run.proto (hunk 3 above)
Comment (darwin67) [reply proto/run/v2/run.proto:56]: `RUN` is the function run itself, `STEP` is the steps within the run.

Comment (darwin67) [reply proto/run/v2/run.proto:56]: `RUN` is the function run itself, `STEP` is the steps within the run.

File: cmd/commands/dev.go (hunk 4 above)
Comment (darwin67) [reply cmd/commands/dev.go:99]: nope it won't. to enable NATS, the `Type` a couple of lines above also need to be changed to `itrace.TracerTypeNATS`.
So just setting this configuration won't have any impact.

File: pkg/pubsub/broker/nats.go (hunk 1 above)
Comment (darwin67) [reply pkg/pubsub/broker/nats.go]: done

File: proto/run/v2/run.proto (hunk 3 above)
Comment (djfarrelly) [reply proto/run/v2/run.proto:56]: Thanks - that was misleading to me.

**Pull request merged**